c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -3296,7 +3298,39 @@ class DeviceCachingAllocator {
       C10_CUDA_CHECK(c10::cuda::SetDevice(stream.device_index()));
 
       EventPool::Event event = create_event_internal(stream.device_index());
-      C10_CUDA_CHECK(cudaEventRecord(*event, stream.stream()));
+      cudaStream_t stream_id = stream.stream();
+
+      if (C10_LIKELY(captures_underway.empty())) {
+        C10_CUDA_CHECK(cudaEventRecordWithFlags(
+            *event, stream_id, cudaEventRecordDefault));
+      } else {
+        // During graph capture, determine if this stream is participating in
+        // any active capture
+        cudaStreamCaptureStatus status{};
+        C10_CUDA_CHECK(cudaStreamGetCaptureInfo(stream_id, &status, nullptr));
+        bool stream_is_capturing = false;
+        for (auto& entry : captures_underway) {
+          if (entry.second(stream_id)) {
+            // Stream is part of an active graph capture - use external
+            // recording to ensure the event is properly included in the graph
+            stream_is_capturing = true;
+            C10_CUDA_CHECK(cudaEventRecordWithFlags(
+                *event, stream_id, cudaEventRecordExternal));
```

Comment:
what do you think marking event as external is achieving here compared to just regular event? You are not using it outside of graph, which is a sole purpose of external event.